cudax/include/cuda/experimental/__async/continue_on.cuh

```diff
@@ -0,0 +1,271 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "completion_signatures.cuh"
+#include "cpos.cuh"
+#include "exception.cuh"
+#include "meta.cuh"
+#include "queries.cuh"
+#include "tuple.cuh"
+#include "utility.cuh"
+#include "variant.cuh"
+
+// This must be the last #include
+#include "prologue.cuh"
+
+namespace cuda::experimental::__async
+{
+struct continue_on_t
+{
+#ifndef __CUDACC__
+
+private:
+#endif
+  template <class... As>
+  using _set_value_tuple_t = _tuple<set_value_t, _decay_t<As>...>;
+
+  template <class Error>
+  using _set_error_tuple_t = _tuple<set_error_t, _decay_t<Error>>;
+
+  using _set_stopped_tuple_t = _tuple<set_stopped_t>;
+
+  using _complete_fn = void (*)(void*) noexcept;
+
+  template <class... Ts>
+  using _set_value_completion =
+    _mif<_nothrow_decay_copyable<Ts...>,
+         completion_signatures<set_value_t(_decay_t<Ts>...)>,
+         completion_signatures<set_value_t(_decay_t<Ts>...), set_error_t(::std::exception_ptr)>>;
+
+  template <class Error>
+  using _set_error_completion =
+    _mif<_nothrow_decay_copyable<Error>,
+         completion_signatures<set_error_t(_decay_t<Error>)>,
+         completion_signatures<set_error_t(_decay_t<Error>), set_error_t(::std::exception_ptr)>>;
+
+  template <class Rcvr, class Result>
+  struct _rcvr_t
+  {
+    using receiver_concept = receiver_t;
+    Rcvr _rcvr;
+    Result _result;
+    _complete_fn _complete;
+
+    template <class Tag, class... As>
+    _CCCL_HOST_DEVICE void operator()(Tag, As&... as) noexcept
+    {
+      Tag()(static_cast<Rcvr&&>(_rcvr), static_cast<As&&>(as)...);
+    }
+
+    template <class Tag, class... As>
+    _CCCL_HOST_DEVICE void _set_result(Tag, As&&... as) noexcept
+    {
+      using _tupl_t = _tuple<Tag, _decay_t<As>...>;
+      if constexpr (_nothrow_decay_copyable<As...>)
+      {
+        _result.template emplace<_tupl_t>(Tag(), static_cast<As&&>(as)...);
+      }
+      else
+      {
+        _CUDAX_TRY( //
+          ({ //
+            _result.template emplace<_tupl_t>(Tag(), static_cast<As&&>(as)...);
+          }),
+          _CUDAX_CATCH(...)( //
+            { //
+              __async::set_error(static_cast<Rcvr&&>(_rcvr), ::std::current_exception());
+            }))
```

Comment:
No change requested. We moved towards exception guards rather than try catch blocks, would that be applicable here

Reply:
> exception guards

do you mean scope guards? that's not applicable here. this code must catch and handle the exception.

Reply:
no, because it would let the exception propagate. this code needs to catch and handle the exception.